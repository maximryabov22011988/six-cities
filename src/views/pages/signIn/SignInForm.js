import React from 'react';
import PropTypes from 'prop-types';

import Form from '../../components/Form';
import Field from '../../components/Field';
import Button from '../../components/Button';

const propTypes = {
  onSignIn: PropTypes.func.isRequired
};

const fieldClasses = {
  wrap: 'login__input-wrapper',
  label: 'visually-hidden',
  input: 'login__input'
};

const field = {
  email: {
    label: 'E-mail',
    type: 'email',
    maxLength: 120,
    name: 'email',
    placeholder: 'Email',
    required: true
  },
  password: {
    label: 'Password',
    type: 'password',
    minLength: 3,
    name: 'password',
    placeholder: 'Password',
    required: true
  }
};

const emailRef = React.createRef();

class SignInForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      email: '',
      password: ''
    };
  }

  componentDidMount() {
    emailRef.current.focus();
  }

  handleChange = evt => {
    const { name, value } = evt.target;
    this.setState({
      [name]: value
    });
  };

  handleSubmit = evt => {
    const { onSignIn } = this.props;
    evt.preventDefault();
    const user = new FormData(evt.target);
    onSignIn(user.get('email'), user.get('password'));
  };

  isLockedButton() {
    const { email, password } = this.state;
    return !(email.length >= 5 && password.length >= 3);
  }

  render() {
    const { email, password } = this.state;

    return (
      <React.Fragment>
        <h1 className="login__title">Sign in</h1>
        <Form
          className="login__form"
          method="post"
          onSubmit={this.handleSubmit}
        >
          <Field
            {...field.email}
            ref={emailRef}
            classes={fieldClasses}
            value={email.value}
            onChange={this.handleChange}
          />
          <Field
            {...field.password}
            classes={fieldClasses}
            value={password.value}
            onChange={this.handleChange}
          />
          <Button
            className="login__submit form__submit"
            type="submit"
            disabled={this.isLockedButton()}
          >
            Sign in
          </Button>
        </Form>
      </React.Fragment>
    );
  }
}

SignInForm.propTypes = propTypes;

export default SignInForm;
