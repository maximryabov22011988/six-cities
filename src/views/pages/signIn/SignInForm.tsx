import * as React from 'react';

import Form from '../../components/Form';
import Field from '../../components/Field';
import Button from '../../components/Button';

import { onSignIn } from '../../types';

const fieldClasses = {
  wrap: 'login__input-wrapper',
  label: 'visually-hidden',
  input: 'login__input',
};

const field = {
  email: {
    label: 'E-mail',
    type: 'email',
    maxLength: 120,
    name: 'email',
    placeholder: 'Email',
    required: true,
  },
  password: {
    label: 'Password',
    type: 'password',
    minLength: 3,
    name: 'password',
    placeholder: 'Password',
    required: true,
  },
};

interface Props {
  onSignIn: onSignIn;
}

interface State {
  email: string;
  password: string;
}

class SignInForm extends React.Component<Props, State> {
  emailRef = React.createRef<HTMLInputElement>();

  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
    };
  }

  componentDidMount() {
    const emailInput = this.emailRef.current;
    if (emailInput) {
      emailInput.focus();
    }
  }

  handleChange = (evt) => {
    const { name, value } = evt.target;
    this.setState({
      [name]: value,
    } as Pick<State, keyof State>);
  };

  handleSubmit = (evt) => {
    const { onSignIn } = this.props;
    evt.preventDefault();
    const user = new FormData(evt.target);
    onSignIn(user.get('email'), user.get('password'));
  };

  isLocked() {
    const { email, password } = this.state;
    return !(email.length >= 5 && password.length >= 3);
  }

  render() {
    const { email, password } = this.state;

    return (
      <React.Fragment>
        <h1 className="login__title">Sign in</h1>
        <Form className="login__form" method="post" onSubmit={this.handleSubmit}>
          <Field
            {...field.email}
            classes={fieldClasses}
            ref={this.emailRef}
            value={email}
            onChange={this.handleChange}
          />
          <Field {...field.password} classes={fieldClasses} value={password} onChange={this.handleChange} />
          <Button className="login__submit form__submit" disabled={this.isLocked()} type="submit">
            Sign in
          </Button>
        </Form>
      </React.Fragment>
    );
  }
}

export default SignInForm;
