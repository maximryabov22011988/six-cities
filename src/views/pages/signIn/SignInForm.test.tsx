import * as React from 'react';
import * as renderer from 'react-test-renderer';

import SignInForm from './SignInForm';

it('SignInForm renders correctly with mandatory props', () => {
  const tree = renderer.create(<SignInForm onSignIn={jest.fn()} />).toJSON();
  expect(tree).toMatchSnapshot();
});
