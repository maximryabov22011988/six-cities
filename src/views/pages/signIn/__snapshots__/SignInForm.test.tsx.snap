// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm renders correctly with mandatory props 1`] = `
Array [
  <h1
    className="login__title"
  >
    Sign in
  </h1>,
  <form
    className="login__form form"
    method="post"
    onSubmit={[Function]}
  >
    <div
      className="login__input-wrapper form__input-wrapper"
    >
      <label
        className="visually-hidden form__label"
      >
        E-mail
      </label>
      <input
        className="login__input form__input"
        maxLength={120}
        name="email"
        onChange={[Function]}
        placeholder="Email"
        required={true}
        type="email"
        value=""
      />
    </div>
    <div
      className="login__input-wrapper form__input-wrapper"
    >
      <label
        className="visually-hidden form__label"
      >
        Password
      </label>
      <input
        className="login__input form__input"
        minLength={3}
        name="password"
        onChange={[Function]}
        placeholder="Password"
        required={true}
        type="password"
        value=""
      />
    </div>
    <button
      className="login__submit form__submit button"
      disabled={true}
      type="submit"
    >
      Sign in
    </button>
  </form>,
]
`;
